lib/pbench/test/unit/server/conftest.py

```diff
@@ -164,19 +164,21 @@ def server_config_shared_db(pytestconfig, monkeypatch):
         a PbenchServerConfig object the test case can use
     """
     cfg_file = pytestconfig.cache.get("_PBENCH_SERVER_CONFIG", None)
+    shutil.copyfile(cfg_file, f"{cfg_file}.shared_db")
     tmp_db = Path(pytestconfig.cache.get("TMP", None)) / "tmp.db"
 
-    with fileinput.FileInput(cfg_file, inplace=True, backup=".bak") as file:
+    with fileinput.FileInput(
+        f"{cfg_file}.shared_db", inplace=True, backup=".bak"
+    ) as file:
         for line in file:
             print(
                 line.replace(
                     "db_uri = sqlite:///:memory:", f"db_uri = sqlite:///{tmp_db}"
                 ),
-                end=" ",
             )
-    monkeypatch.setenv("_PBENCH_SERVER_CONFIG", cfg_file)
+    monkeypatch.setenv("_PBENCH_SERVER_CONFIG", f"{cfg_file}.shared_db")
 
-    return get_server_config()
+    yield get_server_config()
```

Comment:
I'm not sure this adds much since you're not yielding inside a context nor doing anything afterwards. The test case will return here only to immediately return again from the fixture method. Does that buy us anything?

Also, because the `setup` fixture is `scope="session"` won't this hackery persist across test functions unless explicitly undone?

Reply:
Yeah in this context I think yield is just same as return but what do you mean by undoing the fixture? you mean using something like `scope="function"` for this `@pytest.fixture `

Reply:
You've created a new config file, and you're returning a `PbenchServerConfig` from that file. Granted, the default scope of the `monkeypatch` I think is the unit test case, but the file remains, and that bothers me.

I think that, ideally, you'd make this a temp file (like `setup` does with the original), and `yield` to the test inside a `with` context both for that temp file lifetime and the monkeypatch: that way, it becomes clear that returning from the test case will undo the patch and remove the temp file without leaving any mess for the next test.

Reply:
@dbutenhof, what makes you think this fixture is `scope="session"`?  (Is that the default?...or, is that the default for imported modules?  I thought the [default](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) was `scope="function"`.)

The results of fixtures are supposedly cached, so the file copy operation should only happen once per session, I believe.  And, it's a temporary file, so I expect that PyTest will remove it at the end of the session.

But, there _is_ the issue of the patched value of `_PBENCH_SERVER_CONFIG`:  I expect that that will stay patched to point at the `{cfg_file}.shared_db` once this fixture is invoked, until it's overridden by something...and that could be a problem.

So, arguably, we should keep the `yield` and add code after it to restore the `_PBENCH_SERVER_CONFIG` value.  But, I don't know how that will work with subsequent invocations of the fixture.  :-p

(If we're not going to have any code after the `yield`, then we should change it back to a `return`.)

Reply:
Unfortunately, the proper solution here is to remove the database from the equation by mocking its methods....

Reply:
The config file is already created in the temp directory so it will get cleaned up after the test session (just like `server_config` pytest fixture) and the subsequent tests that calls this fixture undo the patch anyway by calling the function again. 

> But, there _is_ the issue of the patched value of `_PBENCH_SERVER_CONFIG`: I expect that that will stay patched to point at the `{cfg_file}.shared_db` once this fixture is invoked, until it's overridden by something...and that could be a problem.

@webbnh I am not sure because the subsequent tests which do not invoke this fixture gets the original config file that was set if called with an original server_config fixture.

Reply:
> @dbutenhof, what makes you think this fixture is `scope="session"`? (Is that the default?...or, is that the default for imported modules? I thought the [default](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) was `scope="function"`.)

I read it in the source file: `@pytest.fixture(scope="session", autouse=True)` ;-)

In other words, the config setup is run once per pytest command session, and won't be automatically "fixed" after you break it here. It's not entirely clear to me how your `monkeypatch.setenv("_PBENCH_SERVER_CONFIG", f"{cfg_file}.shared_db")` interacts with the original fixture's`pytestconfig.cache.set("_PBENCH_SERVER_CONFIG", str(cfg_file))`, but the fact that I need to worry about this, and that the answer isn't immediately obvious, bothers me. Explicitly limiting your monkeypatch with a `yield` inside a monkeypatch context would help; although, again, it's not entirely obvious how that retraction interacts since the previous value was set internally by the `pytestconfig.cache.set` rather than as an external environment variable. We can guess that it has a similar effect, but I don't know that for sure.

I strongly feel that we should be handling the sort of inter-command dependencies you're dealing with here via clean mocking rather than by extending the test DB to persist between commands. As previously stated, this is crossing a busy highway between unit testland and functional testland, and it makes me uncomfortable.

And Webb's correct that technically we shouldn't be using a real DB at all for unit tests in the first place. It's a big job to completely retrofit all of this and I'm not suggesting you should do that right now; but I really don't want to see it extended even further. especially relying on these sort of hazily defined interactions between different fixtures of mixed scope.

Reply:
Okay I updated the PR to use monkeypatch to mock User db querying methods for the testing. However, I do feel the functional testing is necessary here to ensure all the bits work correctly as expected but I will add that code in server functional testing directory in a separate PR.

Reply:
> Okay I updated the PR to use monkeypatch to mock User db querying methods for the testing. However, I do feel the functional testing is necessary here to ensure all the bits work correctly as expected but I will add that code in server functional testing directory in a separate PR.

We definitely need a real functional test environment, where we can lob sequences of real requests against a real server. I'm not convinced that the sort of semi-black-box cross-component "functional" unit testing you're talking about here is of much real value... although in the absence of a real functional test pipeline, we could consider it. (But I'd rather put the effort towards a nightly live-hardware functional test environment.)

Reply:
> I read it in the source file

Ah.  Sorry -- @dbutenhof, you were talking about the `setup` fixture and I thought we were talking about the Nikhil's new DB reconfiguration fixture.

> I am not sure because the subsequent tests which do not invoke this fixture gets the original config file that was set if called with an original server_config fixture.

Um, @npalaska, "sure" is not a word I would use regarding any details of PyTest....  :-}  I _presume_ that it works hard to provide a nice layering of fixtures with the effects of per-function fixtures being rolled back at the ends of the functions (and so on), but, if a fixture has an external side-effect (such as creating or modifying a file), I don't see how PyTest on it's own can know to roll those back, and I have no idea whether modifying an environment variable (or modifying a method for accessing an environment variable?) is an _external_ side-effect or not.

So, I think the answer is, let's abandon this craziness and use mocks (which I think you've now done -- I'll go admire your work in a minute).